### PR TITLE
refactor: use abstract UserAddress type for author/voter addresses

### PR DESCRIPTION
**Pull request by Orland0x**
- We switch to a general `UserAddress` enum that can represent both a starknet and ethereum address. this enum can be further extended to include any number of other address types also.
- There is also a custom address type that we represent as a u256. This provides compatibility  for any address type that fits in a u256, without the strongly typed support though.

### CODE REVIEW

File: starknet/src/utils/types.cairo [hunk 1]
```diff
@@ -6,14 +6,88 @@ use hash::LegacyHash;
...
+    }
+}
+
+trait UserAddressTrait {
+    fn to_starknet_address(self: UserAddress) -> ContractAddress;
+    fn to_ethereum_address(self: UserAddress) -> EthAddress;
+    fn to_custom_address(self: UserAddress) -> u256;
+}
+
+impl UserAddressImpl of UserAddressTrait {
```
Comment (Orland0x) [inline starknet/src/utils/types.cairo]: not sure whether theres an easier way to do this

File: .gitignore [hunk 2]
```diff
@@ -3,11 +3,12 @@ node_modules/
...
 
+build/
 cache/
 out/
 
 !/broadcast
 /broadcast/*/31337/
 /broadcast/**/dry-run/
 
-docs/
```
Comment (pscott) [inline .gitignore:13]: sadge

File: starknet/src/utils/types.cairo [hunk 3]
```diff
@@ -6,14 +6,88 @@ use hash::LegacyHash;
...
 
+#[derive(Copy, Drop, Serde, LegacyHash, PartialEq, starknet::Store)]
+enum UserAddress {
+    // Starknet address type
+    StarknetAddress: ContractAddress,
+    // Ethereum address type
+    EthereumAddress: EthAddress,
+    // Custom address type to provide compatibility with any address that can be represented as a u256.
+    CustomAddress: u256
+}
```
Comment (pscott) [inline starknet/src/utils/types.cairo]: Quick naming comment: the word address is always repeated, seems a bit redundant.

How about:
```
enum Address {
  Starknet: ContractAddress,
  Ethereum: EthAddress,
  Custom: u256,
}
```

File: starknet/src/utils/types.cairo [hunk 4]
```diff
@@ -6,14 +6,88 @@ use hash::LegacyHash;
...
 use array::SpanTrait;
 
+#[derive(Copy, Drop, Serde, LegacyHash, PartialEq, starknet::Store)]
+enum UserAddress {
+    // Starknet address type
+    StarknetAddress: ContractAddress,
+    // Ethereum address type
+    EthereumAddress: EthAddress,
+    // Custom address type to provide compatibility with any address that can be represented as a u256.
+    CustomAddress: u256
```
Comment (pscott) [inline starknet/src/utils/types.cairo]: Should the customAddress be an `Array::<felt252>` ? something that devs could decode?

File: starknet/src/utils/types.cairo [hunk 5]
```diff
@@ -6,14 +6,88 @@ use hash::LegacyHash;
...
-    ContractAddress, Store, contract_address_const, StorageBaseAddress, SyscallResult,
+    EthAddress, ContractAddress, Store, contract_address_const, StorageBaseAddress, SyscallResult,
     storage_write_syscall, storage_read_syscall, storage_address_from_base_and_offset,
     storage_base_address_from_felt252, contract_address::Felt252TryIntoContractAddress,
     syscalls::deploy_syscall, class_hash::Felt252TryIntoClassHash
 };
 use sx::utils::math::pow;
 use array::SpanTrait;
 
+#[derive(Copy, Drop, Serde, LegacyHash, PartialEq, starknet::Store)]
```
Comment (pscott) [inline starknet/src/utils/types.cairo]: I believe `LegacyHash` should not be used when possible. Are we sure we wish to derive it?

File: starknet/src/utils/types.cairo [hunk 6]
```diff
@@ -6,14 +6,88 @@ use hash::LegacyHash;
...
+trait UserAddressTrait {
+    fn to_starknet_address(self: UserAddress) -> ContractAddress;
+    fn to_ethereum_address(self: UserAddress) -> EthAddress;
+    fn to_custom_address(self: UserAddress) -> u256;
+}
+
+impl UserAddressImpl of UserAddressTrait {
+    fn to_starknet_address(self: UserAddress) -> ContractAddress {
+        match self {
+            UserAddress::StarknetAddress(address) => address,
```
Comment (pscott) [inline starknet/src/utils/types.cairo]: in Rust you can do :
```
match self {
  UserAddress::StarknetAddress(address) => address,
  _ => panic_with_felt252('Incorrect address type'),
}
```
does it work in cairo?

**Review by pscott — APPROVED** [review]
minor comments but all good!

File: starknet/src/utils/types.cairo (hunk 4 above)
Comment (Orland0x) [reply starknet/src/utils/types.cairo]: dynamic type adds overhead so seems unnecessary i think

File: starknet/src/utils/types.cairo (hunk 5 above)
Comment (Orland0x) [reply starknet/src/utils/types.cairo]: removed

File: starknet/src/utils/types.cairo (hunk 6 above)
Comment (Orland0x) [reply starknet/src/utils/types.cairo]: i dont think so yet. https://book.cairo-lang.org/ch05-02-the-match-control-flow-construct.html#match-0-and-the-_-placeholder

Comment (Orland0x) [reply starknet/src/utils/types.cairo]: `Unsupported match. Currently, matches require one arm per variant, in the order of variant definition.`

Comment (pscott) [reply starknet/src/utils/types.cairo]: sucks, weird that the cairo book doesn't mention that. Will probs come in the future

File: starknet/src/utils/types.cairo (hunk 4 above)
Comment (pscott) [reply starknet/src/utils/types.cairo]: idea behind it is we don't want to be limited to u256 if a new chain uses bigger addresses but yeah I'm fine with that

**Pull request merged**